### PR TITLE
Remove use of deprecated null_data_source

### PR DESCRIPTION
**Pull request by bwmetcalf**
Moved logic from null_data_source into definition of locals.public_dns.

null_data_source is now deprecated (using tf 0.14.5) and throws warnings.  See PR below.

https://registry.terraform.io/providers/hashicorp/null/latest/docs/data-sources/data_source

https://github.com/cloudposse/terraform-aws-ec2-instance/issues/93

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -150,12 +153,6 @@ resource "aws_eip" "default" {
   tags              = module.this.tags
 }
 
-data "null_data_source" "eip" {
-  inputs = {
-    public_dns = "ec2-${replace(join("", aws_eip.default.*.public_ip), ".", "-")}.${local.region == "us-east-1" ? "compute-1" : "${local.region}.compute"}.amazonaws.com"
-  }
-}
-
 resource "aws_ebs_volume" "default" {
```
[the remark this answers is not in the document]
Comment (woz5999) [reply main.tf:156]: this seems to be a false positive? In any case, it's unrelated to the changes in this PR.

**Comment by Gowiem** [comment]
@bwmetcalf our CI got stuck which happens sometimes. Can you push an empty commit like so: 

```bash
git commit --allow-empty -m "Trigger CI"
git push
```

That will trigger that last required build and then @woz5999 can take it from here 👍

**Comment by bwmetcalf** [comment]
@Gowiem Hey Matt, anything else I need to do on this to get it merged?  Thanks.

**Comment by woz5999** [comment]
@Gowiem I'm not sure of the solution to the Bridgecrew analysis failure. It's unrelated to this change and seems to be invalid regardless.

**Comment by Gowiem** [comment]
@woz5999 agreed. Let's get @maximmi or @Nuru thoughts on the matter. They're more involved in the BridgeCrew work than I am.

Gents, do we need to hold off merging this PR? It seems the BridgeCrew CI is failing on this PR due to something that is already merged into master. How are we handling that elsewhere?

**Comment by jwstric2** [comment]
hi team, I hit a similar issue on https://github.com/cloudposse/terraform-aws-alb/pull/86, drew the short straw as a PR that went in after enhancements to code analysis.  it's definitely new changes to the analytics of bridgecrew as I was able to pull an old container and never saw the results produced in the new containers.  Seems the consensus is skips if they make sense to skip ...

File: main.tf (hunk 1 above)
[the remark this answers is not in the document]
Comment (Nuru) [reply main.tf:156]: Yes, don't worry about this bridgecrew warning

File: main.tf [hunk 2]
```diff
@@ -11,7 +11,10 @@ locals {
   ami                    = var.ami != "" ? var.ami : join("", data.aws_ami.default.*.image_id)
   ami_owner              = var.ami != "" ? var.ami_owner : join("", data.aws_ami.default.*.owner_id)
   root_volume_type       = var.root_volume_type != "" ? var.root_volume_type : data.aws_ami.info.root_device_type
-  public_dns             = var.associate_public_ip_address && var.assign_eip_address && module.this.enabled ? data.null_data_source.eip.outputs["public_dns"] : join("", aws_instance.default.*.public_dns)
+  public_dns = (var.associate_public_ip_address && var.assign_eip_address && module.this.enabled ?
+    ("ec2-${replace(join("", aws_eip.default.*.public_ip), ".", "-")}.${local.region == "us-east-1" ?
+    "compute-1" : "${local.region}.compute"}.amazonaws.com") :
+  join("", aws_instance.default.*.public_dns))
```
Comment (Nuru) [inline main.tf:17]: While I appreciate the effort, this is really hard to read. Let's use local variables to simplify the expression.
```suggestion

  region_domain  = local.region == "us-east-1" ? "compute-1.amazonaws.com" : "${local.region}.compute.amazonaws.com"
  eip_public_dns = "ec2-${replace(join("", aws_eip.default.*.public_ip), ".", "-")}.${local.region_domain}"
  public_dns     = (
    var.associate_public_ip_address && var.assign_eip_address && module.this.enabled ?
    local.eip_public_dns : join("", aws_instance.default.*.public_dns)
  )
```

**Review by Nuru — CHANGES_REQUESTED** [review]
Do not worry about the Bridgecrew warning/complaint, but please simplify the `public_dns` expression. I made a suggestion but have not double-checked it for accuracy.

**Review by woz5999 — APPROVED** [review]

**Comment by woz5999** [comment]
@Gowiem I think this is ready to go

**Review by Gowiem — APPROVED** [review]

**Comment by Gowiem** [comment]
Thanks for the contribution @bwmetcalf (and help @jwstric2 + @woz5999)! Released as [0.31.0](https://github.com/cloudposse/terraform-aws-ec2-instance/releases/tag/0.31.0).

**Pull request merged**